### DISCUSSION TITLE
Status of Strimzi with Kraft?

### DISCUSSION THREAD

**Discussion by villesau** [strimzi/strimzi-kafka-operator#8358] [Q&A]
Hi,

It looks like Kraft is production ready at least for new clusters: https://blogs.apache.org/kafka/entry/what-rsquo-s-new-in

I'd like to understand what is the status with Kraft & Strimzi? Is it ready for prime time? Especially when setting things up from scratch.

**Comment by scholzj** [comment]
I guess the production readiness is a question of opinion as there are many missing features today which are quite important for many users such as SCRAM-SHA authentication, JBOD support, upgrades from ZooKeeper based clusters etc. Please search for the previous discussions here and on Slack for more details. 

You can also check the https://github.com/strimzi/proposals/pull/77 proposal and the video linked there for how it might look like when it is ready.

**Reply by villesau** [reply]
Thank you! Is it still generally advised against for using Kafka with Kraft when setting up completely new setup? Or does it depend on scale, or something else? Is there some timeline? For a new user the situation is slightly confusing since different sources gives a bit different view on things.

**Reply by scholzj** [reply]
As for Strimzi, KRaft is not considered production ready. That is also what the documentation says as well IIRC. I'm in no official position to speak for the Apache Kafka project itself. There it is just my personal opinion.

**Reply by lucasfcnunes** [reply]
@scholzj 

Production ready officially in 3.3.1/2022-10-03 (actually in 3.3.0/2022-08, [but](https://kafka.apache.org/downloads#:~:text=A%20significant%20bug%20was%20found%20in%20the%203.3.0))
- https://adrianbednarz.medium.com/kraft-is-production-ready-6492ae9015b0
- [KAFKA/KIP-833](https://cwiki.apache.org/confluence/display/KAFKA/KIP-833%3A+Mark+KRaft+as+Production+Ready#:~:text=every%204%20months.-,Kafka%203.3%20(released),KRaft%20mode%20declared%20production%2Dready,-Kafka%203.4%20(released))

So:
> Kafka 3.3 (released)
August 2022
KRaft mode declared production-ready

> Kafka 4.0
April 2024
Only KRaft mode supported.

**Reply by scholzj** [reply]
Not sure I understand what are you trying to say.

**Reply by lucasfcnunes** [reply]
Oh.

I was addressing the following:
> I'm in no official position to speak for the Apache Kafka project itself. There it is just my personal opinion.

by bringing up the official Apache Kafka project's position on KRaft:
- that it is GA/production ready since 2022-08
- that ~1 year they will start releasing without ZK (Kafka 4.0)

**Reply by scholzj** [reply]
Yeah, I know about all of that. that is why I made it clear that that is my opinion.

**Reply by bygui86** [reply]
@scholzj I see your opinion, but can you please share which features are missing in KRaft to be considered production ready?

**Reply by scholzj** [reply]
Most of them are repeated again and again. The only thing which changed really in 3.5.0 is SCRAM-SHa support. But migration is not production-ready, JBOD support is missing, the KRaft controllers are not really operationalized (no Admin API access / management => so it is really hard to operate them).

**Comment by vyom-soft** [comment]
Hello Scholzj, 
We are on Strimzi 0.36 with default zookeeper. We would like to use Kraft. If I enable the UseKRaft feature 
gate, would that be sufficient. ?  To get rid of Zookeeper. Will the kafka-cluster work as usual ?
Can you please recommend how to switch to KRaft without having to uninstall the running strimzi-kafka-operater and without destroying the existing kafka cluster running on other NS which the operator monitors?

**Reply by ppatierno** [reply]
It's not possible yet. Because ... first, the KRaft support in Strimzi is still behind a feature gate and it's not production ready so you can use it just for testing/development, not from production cluster and only for creating a cluster from scratch. Second, the migration process is under development. There was a [proposal](https://github.com/strimzi/proposals/blob/main/059-zk-kraft-migration.md) approved about the flow but the implementation is not ready yet.

**Reply by ppatierno** [reply]
Just adding that everything is well stated in the official doc here https://strimzi.io/docs/operators/latest/deploying#ref-operator-use-kraft-feature-gate-str

**Reply by olucha** [reply]
Hello @ppatierno @scholzj ,
I'm evaluating Strimzi for a use case, and Kraft support is one the requisites. With current Kraft support status Strimzi is a no-go, but taking into account current limitations, the one hurting us the most is the impossibility to upgrade both Strimzi operator and Kafka brokers.
I have tried to find in project roadmap an estimation for the needed developments that would address those limitations, without luck.
I have around 6 months to evaluate the final architecture of the solution, and I could start with Strimzi if it's clear that this time window would be enough to have, at least, that upgrade support in place.
Please, can you give me a hint on the expected or planned evolution of the UseKraft feature gate?
Thanks a lot!

**Reply by bygui86** [reply]
@olucha sorry but your question is not relevant to this stream, please remove it and put it in the Slack channel or in another GitHub discussion

**Reply by scholzj** [reply]
@olucha I do not have any timeline for Strimzi. If you check our PRs, you will see that we are actively working on it. And if you look at the Kafka mailing lists, it seems clear that Kafka does not really have a timeline either. Keep in mind that Apache Kafka is still missing many important features as well.

> I have around 6 months to evaluate the final architecture of the solution, and I could start with Strimzi if it's clear that this time window would be enough to have, at least, that upgrade support in place.

I would expect us to have the missing features in Strimzi earlier than in 6 months. But I'm not sure I would consider it production-ready by then neither in Strimzi nor in Kafka. 

But let me ask you ... I do not know what solution are you building ... but what KRaft features do you require for it? What are the benefits KRaft gives you? Why do you so desperately need KRaft for it? Do you run any KRaft-based Kafka clusters in production to understand the limitations? In the end, it is always your decision and your risks to take of course. But insisting on deciding on this right now regardless the quality makes me curious.

**Reply by olucha** [reply]
Sorry @bygui86, I thought asking about the roadmap for the support of UseKraft feature gate was related with a discussion called "Status of Strimzi with Kraft". No problem to delete the question and redirect it to Slack.

**Reply by ppatierno** [reply]
@olucha I would keep this here tbh. It's related to the status, and now we are curious about the questions Jakub asked to you :-)

**Reply by olucha** [reply]
Hello @scholzj , and thanks for your answer.
I don't actively follow Kafka's development or mailing list, I just know the "official statement" about readiness of Kraft from both Apache and Confluent, although I have already read your previous answers about your concerns to the current degree of completion of Kraft support en Apache Kafka itself.
I have to recognize that it's hard to get the complete picture when there's a KIP-500 with an associated Jira ticket marked as complete, but a lot of other KIPs that span from that... that's my fault.
Returning to your questions, I'm not desperate to have Kraft in place :-), but my client sees it as an improvement over current setup with Zookeeper+Kafka, and wants to architect a solution based on it. This is a hard requirement not stated by me...
I have seen your PR https://github.com/strimzi/strimzi-kafka-operator/pull/9364, but looking to https://github.com/strimzi/proposals/blob/main/061-kraft-upgrades-and-downgrades.md I'm not sure if there are too much things to do before that particular functionality, upgrades and downgrades, be complete.

**Reply by scholzj** [reply]
> I have to recognize that it's hard to get the complete picture when there's a KIP-500 with an associated Jira ticket marked as complete, but a lot of other KIPs that span from that... that's my fault.

Yeah, unfortunately, there are many additional KIPs since KIP-500 dealing with various aspects of KRaft. So it is quite hard to track. Plus even the KIPs don't give you a complete picture. There are also bugs in the code. And there is also a lack of APIs for operationalizing (automating) the KRaft controllers or the migration process -> which is what slows us down because these things are complicated to work around.

> Returning to your questions, I'm not desperate to have Kraft in place :-), but my client sees it as an improvement over current setup with Zookeeper+Kafka, and wants to architect a solution based on it. This is a hard requirement not stated by me...

I think it would still be good to understand the reasons. I see a lot of misunderstanding among users about what will it actually bring. Don't get me wrong - I think getting rid of ZooKeeper is not a bad thing and I will not shed a tear when it is gone. But the reality for most Strimzi users is that in production the ZooKeeper pods will be replaced by the KRaft controller pods and that is it. There might be some improvements to how much Kafka can scale. But most users do not run at the scale to really utilize them. So it would be still interesting to understand why they require KRaft.

> I have seen your PR https://github.com/strimzi/strimzi-kafka-operator/pull/9364, but looking to https://github.com/strimzi/proposals/blob/main/061-kraft-upgrades-and-downgrades.md I'm not sure if there are too much things to do before that particular functionality, upgrades and downgrades, be complete.

#9364 is the first part of the upgrades. There will be a second part. as well soon I currently expect that to land in the next Strimzi version. But there is also #9324 for the migration, #9146 (that is hopefully close to being merged as well - and is essential for the upgrades as well), etc. So there is more work in progress.
